component/metrics/remotewrite/remote_write.go

```diff
@@ -0,0 +1,289 @@
+			// inactive and scheduled for deletion.
+			ts := c.remoteStore.LowestSentTimestamp() - minWALTime.Milliseconds()
+			if ts < 0 {
+				ts = 0
+			}
+
+			// Network issues can prevent the result of getRemoteWriteTimestamp from
+			// changing. We don't want data in the WAL to grow forever, so we set a cap
+			// on the maximum age data can be. If our ts is older than this cutoff point,
+			// we'll shift it forward to start deleting very stale data.
+			if maxTS := timestamp.FromTime(time.Now().Add(-maxWALTime)); ts < maxTS {
+				ts = maxTS
+			}
+
+			if ts == lastTs {
+				level.Debug(c.log).Log("msg", "not truncating the WAL, remote_write timestamp is unchanged", "ts", ts)
+				continue
+			}
+			lastTs = ts
+
+			level.Debug(c.log).Log("msg", "truncating the WAL", "ts", ts)
+			err := c.walStore.Truncate(ts)
+			if err != nil {
+				// The only issue here is larger disk usage and a greater replay time,
+				// so we'll only log this as a warning.
+				level.Warn(c.log).Log("msg", "could not truncate WAL", "err", err)
+			}
+		}
+	}
+}
+
+// Update implements Component.
+func (c *Component) Update(newConfig component.Arguments) error {
+	cfg := newConfig.(RemoteConfig)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	var rwConfigs []*config.RemoteWriteConfig
+	for _, rw := range cfg.RemoteWrite {
+		parsedURL, err := url.Parse(rw.URL)
+		if err != nil {
+			return fmt.Errorf("cannot parse remote_write url %q: %w", rw.URL, err)
+		}
+
+		rwc := &config.RemoteWriteConfig{
+			Name:          rw.Name,
+			URL:           &common.URL{URL: parsedURL},
+			RemoteTimeout: model.Duration(30 * time.Second),
+			QueueConfig:   config.DefaultQueueConfig,
+			MetadataConfig: config.MetadataConfig{
+				Send: false,
+			},
+			HTTPClientConfig: common.DefaultHTTPClientConfig,
+		}
+
+		if rw.BasicAuth != nil {
+			rwc.HTTPClientConfig.BasicAuth = &common.BasicAuth{
+				Username: rw.BasicAuth.Username,
+				Password: common.Secret(rw.BasicAuth.Password),
+			}
+		}
+
+		rwConfigs = append(rwConfigs, rwc)
+	}
+
+	err := c.remoteStore.ApplyConfig(&config.Config{
+		GlobalConfig: config.GlobalConfig{
+			ExternalLabels: toLabels(cfg.ExternalLabels),
+		},
+		RemoteWriteConfigs: rwConfigs,
+	})
+	if err != nil {
+		return err
+	}
+
+	c.cfg = cfg
+	return nil
+}
+
+// Receive implements the receiver.receive func that allows an array of metrics to be passed
+func (c *Component) Receive(ts int64, metricArr []*metrics.FlowMetric) {
+	app := c.walStore.Appender(context.Background())
+	for _, m := range metricArr {
+		// TODO this should all be simplified into one call
+		if m.GlobalRefID == 0 {
+			globalID := metrics.GlobalRefMapping.CreateGlobalRefID(m.Labels)
+			m.GlobalRefID = globalID
+		}
+		localID := metrics.GlobalRefMapping.GetLocalRefID(c.opts.ID, m.GlobalRefID)
+		newLocal, err := app.Append(storage.SeriesRef(localID), m.Labels, ts, m.Value)
+		if localID == 0 {
+			metrics.GlobalRefMapping.GetOrAddLink(c.opts.ID, uint64(newLocal), m.Labels)
```

Comment:
The Appender implementation here passed from other components will most likely be Prometheus' [headAppender](https://github.com/prometheus/prometheus/blob/f389711943daca716675fdd3b37901f4d70100f7/tsdb/head_append.go#L244-L301).

Do we also want to check that the returned refID is non-zero (i.e. the Append succeeded) before adding the link?

Reply:
Thats a good call